nexus/mgs-updates/src/rot_bootloader_updater.rs

```diff
@@ -6,35 +6,375 @@
+                            FoundVersion::MissingVersion
+                        } else {
+                            return Err(PrecheckError::from(error));
+                        }
+                    }
+                };
+            match (&expected_stage0_next_version, &found_stage0_next_version) {
+                // expected garbage, found garbage
+                (
+                    ExpectedVersion::NoValidVersion,
+                    FoundVersion::MissingVersion,
+                ) => (),
+                // expected a specific version and found it
+                (
+                    ExpectedVersion::Version(artifact_version),
+                    FoundVersion::Version(found_stage0_next_version),
+                ) if artifact_version.to_string()
+                    == *found_stage0_next_version =>
+                {
+                    ()
+                }
+                // anything else is a mismatch
+                (ExpectedVersion::NoValidVersion, FoundVersion::Version(_))
+                | (ExpectedVersion::Version(_), FoundVersion::MissingVersion)
+                | (ExpectedVersion::Version(_), FoundVersion::Version(_)) => {
+                    return Err(PrecheckError::WrongInactiveVersion {
+                        expected: expected_stage0_next_version.clone(),
+                        found: found_stage0_next_version,
+                    });
+                }
+            };
+
+            // The status is only considered ready for update if the stage0_next
+            // version found in the caboose is valid, and it matches what we
+            // found in the stage_0 caboose. Otherwise, we are waiting for an
+            // ongoing update.
+            match found_stage0_next_version {
+                FoundVersion::Version(v) => {
+                    if v == found_stage0_version {
+                        Ok(PrecheckStatus::ReadyForUpdate)
+                    } else {
+                        Ok(PrecheckStatus::WaitingForOngoingUpdate)
+                    }
+                }
+                FoundVersion::MissingVersion => {
+                    Ok(PrecheckStatus::WaitingForOngoingUpdate)
+                }
+            }
+        }
+        .boxed()
     }
 
     /// Attempts once to perform any post-update actions (e.g., reset the
     /// device)
     fn post_update<'a>(
         &'a self,
-        _log: &'a slog::Logger,
-        _mgs_clients: &'a mut MgsClients,
-        _update: &'a PendingMgsUpdate,
-    ) -> BoxFuture<'a, Result<(), GatewayClientError>> {
-        // TODO-K: To be completed in a follow up PR
-        todo!()
+        log: &'a slog::Logger,
+        mgs_clients: &'a mut MgsClients,
+        update: &'a PendingMgsUpdate,
+    ) -> BoxFuture<'a, Result<(), PostUpdateError>> {
+        async move {
+            // Before setting stage0 to the new version we want to ensure
+            // the image is good and we're not going to brick the device.
```

Comment:
It doesn't seem like we should be able to brick an RoT in any way by sending normal API commands in some kind of bad order. (This is ignoring cases like "lost power at just the wrong time" as described below, since that's a pretty extenuating circumstance.)

If we didn't do this check and the image wasn't good, would it actually brick it?

Reply:
Yeah if my understanding is right, I think it's more like:

"To protect against bricking itself, the device will only activate a new image after it's been verified.  Images are only verified at device boot time.  Thus, we'll reset the device once to cause the signature to be verified.  Then we can activate the new image and reset the device again."

Reply:
Yeah, I meant what @davepacheco said. Clearly my English needs improving 😄